submission_offsets.py

```diff
@@ -0,0 +1,68 @@
+from flask.cli import FlaskGroup
+import click
+from collections import defaultdict
+
+import db
+import webserver
+
+from sqlalchemy import text
+
+cli = FlaskGroup(add_default_commands=False, create_app=webserver.create_app_flaskgroup)
+
+@cli.command(name='add-offsets')
+@click.option("--limit", "-l", default=10000)
+def add_offsets(limit):
+	"""Update lowlevel submission offsets with a specified limit."""
+	incremental_add_offset(limit)
+
+def incremental_add_offset(limit):
+	with db.engine.connect() as connection:
+		# Find the next batch of items to update
+		batch_query = text("""
+			SELECT id, gid
+			  FROM lowlevel
+			 WHERE submission_offset IS NULL
+			 LIMIT :limit
+		""")
+		batch_result = connection.execute(batch_query, { "limit": limit })
+
+		# Find max existing offsets
+		offset_query = text("""
+			SELECT gid, MAX(submission_offset)
+			  FROM lowlevel
+			 WHERE submission_offset IS NOT NULL
+		      GROUP BY gid
+		""")
+		offset_result = connection.execute(offset_query)
+
+		max_offsets = defaultdict(int)
+		for gid, max_offset in offset_result.fetchall():
+			max_offsets[gid] = max_offset
+
+		print("Starting batch insertions...")
+		print("============================")
+		count = 0
+		for id, gid in batch_result.fetchall():
```

Comment:
@alastair When you said that the loop should be within a transaction, is it sufficient for the entire operation to be within the same transaction (starting with `with db.engine.connect() as connection:` on line 19)? Or should I embed another transaction here that is just for the loop?

Reply:
this loop should be in an additional transaction - you can do this with `with connection.begin() as transaction:`